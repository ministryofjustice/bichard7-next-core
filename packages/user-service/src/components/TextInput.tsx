interface Props {
  label?: string
  name: string
  hint?: string
  className?: string
  error?: string | boolean
  width?: "5" | "10" | "20" | "30"
  id?: string
  type?: string
  readOnly?: boolean
  disabled?: boolean
  mandatory?: boolean
  value?: string
  optionalProps?: object
}

const TextInput = ({
  label,
  hint,
  error,
  className,
  name,
  width,
  id,
  readOnly = false,
  disabled = false,
  type = "text",
  value,
  mandatory = false,
  optionalProps = {}
}: Props) => {
  const hintElementId = `${name}-hint`
  const errorElementId = `${name}-error`
  const widthClassName = width ? ` govuk-input--width-${width}` : ""
  const hasError = error && !readOnly

  return (
    <div className={`govuk-form-group${className ? ` ${className}` : ""}${hasError ? " govuk-form-group--error" : ""}`}>
      {label && (
        <label className="govuk-label" htmlFor={name}>
          {label}
          {mandatory && " *"}
        </label>
      )}

      {hint && (
        <div id={hintElementId} className="govuk-hint">
          {hint}
        </div>
      )}
      {!!hasError && error !== true && (
        <span data-test={`text-input_${name}-error`} id={errorElementId} className="govuk-error-message">
          <span className="govuk-visually-hidden">{"Error:"}</span> {error}
        </span>
      )}
      <input
        className={`govuk-input${widthClassName}${hasError ? " govuk-input--error" : ""}`}
        id={id || name}
        name={name}
        data-test={`text-input_${name}`}
        type={type}
        defaultValue={value}
        readOnly={readOnly}
        disabled={disabled}
        aria-describedby={`${hint ? hintElementId : ""} ${hasError ? errorElementId : ""}`}
        {...optionalProps}
      />
    </div>
  )
}

export default TextInput
