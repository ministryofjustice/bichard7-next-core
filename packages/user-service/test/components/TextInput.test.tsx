import { render } from "@testing-library/react"
import TextInput from "components/TextInput"

it("should render text input component", () => {
  const { container } = render(
    <TextInput
      id="inputId"
      name="inputName"
      readOnly
      disabled
      value="Input Value"
      error="Error message"
      label="Input label"
      className="custom-class-name"
      mandatory
      type="email"
      hint="Input hint"
      width="5"
    />
  )

  expect(container).toMatchSnapshot()
})
