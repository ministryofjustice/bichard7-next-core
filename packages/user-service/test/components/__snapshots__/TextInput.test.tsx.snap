// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`should render text input component 1`] = `
<div>
  <div
    class="govuk-form-group custom-class-name"
  >
    <label
      class="govuk-label"
      for="inputName"
    >
      Input label
       *
    </label>
    <div
      class="govuk-hint"
      id="inputName-hint"
    >
      Input hint
    </div>
    <input
      aria-describedby="inputName-hint "
      class="govuk-input govuk-input--width-5"
      data-test="text-input_inputName"
      disabled=""
      id="inputId"
      name="inputName"
      readonly=""
      type="email"
      value="Input Value"
    />
  </div>
</div>
`;
